feat: Add new methods for handling URLs

rroig-at-wiris opened the pull request.

rroig-at-wiris committed:
feat: Add new methods for handling URLs

Add auxiliary methods such as clean_component and slash_lfill

Clean components in initializer and classmethods that receive any URL component

Add new testcases to verify components are cleaned as expected

Add tests for new methods

dmas-at-wiris commented on santoku/utils/test/test_url_handler.py:
```diff
@@ -26,16 +27,16 @@ def test_string_subdomain_is_processed_as_expected(input_subdomain, expected_sub
         (
             "https://sub1.example.com/path/",
-            ("https", ["sub1"], "example", "com", "/path/"),
+            ("https", ["sub1"], "example", "com", "path"),
```
Maybe add a case with uppercases in path?

dmas-at-wiris commented on santoku/utils/test/test_url_handler.py:
```diff
@@ -26,16 +27,16 @@ def test_string_subdomain_is_processed_as_expected(input_subdomain, expected_sub
         ("http://sub2.sub1.example.es", ("http", ["sub2", "sub1"], "example", "es", None)),
         ("https://sub1.example.co.uk", ("https", ["sub1"], "example", "co.uk", None)),
         ("https://example.com", ("https", [], "example", "com", None)),
-        ("https://example.XN--1QQW23A", ("https", [], "example", "XN--1QQW23A", None)),
```
Could be useful to add an explanation of what a `XN--` domain is

dmas-at-wiris commented on santoku/utils/url_handler.py:
```diff
@@ -202,14 +213,69 @@ def process_path(cls, parsed_url: ParseResult) -> Optional[str]:
+        return cls.clean_component(component=parsed_url.path, lowercase=False)
+
+    @classmethod
+    def slash_lfill(cls, component: Optional[str]) -> Optional[str]:
```
Calling this functionality `slash_lfill(...)` is inaccurate imho. fill or zfill tends to be used to pad a string with certain amount of symbols. I'd suggest to call this `prepend_slash(...)`.

dmas-at-wiris commented on santoku/utils/test/test_url_handler.py:
```diff
@@ -212,10 +312,73 @@ def test_exception_is_raised_for_num_subdomain_smaller_than_0(input_url, num_sub
     argvalues=(
         (["example.com"], ["here"], ["example.com", "example.com/here"]),
+        (["example.com"], ["/here"], ["example.com", "example.com/here"]),
+        (["EXAMPLE.com"], ["/HEre"], ["example.com", "example.com/HEre"]),
```
I saw this now :)

dmas-at-wiris commented on santoku/utils/url_handler.py:
```diff
@@ -41,6 +41,8 @@ class URLHandler:
     - If you provide an invalid URL, you might get unexpected results for some of the methods.
```
This comment is linked to the previous line. Can you include a link to the standard specification and explain the differences?

dmas-at-wiris commented on santoku/utils/url_handler.py:
```diff
@@ -41,6 +41,8 @@ class URLHandler:
     - If you provide an invalid URL, you might get unexpected results for some of the methods.
     - When a non-valid URL such as 'fakedomain' is provided, it is considered as `base_domain`.
```
```suggestion
    - When a non-valid URL such as 'fakedomain' is provided, 'fakedomain' will be considered as the `base_domain`.
```

dmas-at-wiris commented on santoku/utils/url_handler.py:
```diff
@@ -202,14 +213,69 @@ def process_path(cls, parsed_url: ParseResult) -> Optional[str]:
+            if not component.startswith("/"):
+                component = f"/{component}"
+
+        return component or None
```
```suggestion
        if not component:
            return None
        if component.startswith("/"):
            return component 
            
         return f"/{component}"
```

dmas-at-wiris commented on santoku/utils/url_handler.py:
```diff
@@ -202,14 +213,69 @@ def process_path(cls, parsed_url: ParseResult) -> Optional[str]:
+            if lowercase:
+                component = component.lower()
+
+        return component or None
```
Maybe use same pattern as with previous function?

dmas-at-wiris commented on santoku/utils/url_handler.py:
```diff
@@ -287,15 +328,129 @@ def join_base_domain_with_suffix(
+        Optional[str]
+            The joined `domain` with `path`. Returns `domain` or `path` if one of them is `None`, or
+            `None` if both are `None`.
+
```
An example could be useful

dmas-at-wiris commented on santoku/utils/url_handler.py:
🤯 
Don't use list comprehension in this case. It's hard to tell what is going on.

hqiu-at-wiris committed:
refactor: improve URLHandler

add more explanations to URLHandler docstring

rename slash_lfill to prepend_slash

make explode_path more readable

dmas-at-wiris commented on santoku/utils/url_handler.py:
```diff
@@ -202,14 +236,70 @@ def process_path(cls, parsed_url: ParseResult) -> Optional[str]:
+            component is empty.
+
+        """
+        if component:
+            component = component.strip("/")
+            if lowercase:
+                component = component.lower()
+
+        return component or None
```
Review this part. `component` is an input parameter

hqiu-at-wiris committed:
perf: improve clean_component method in URL Handler

do the check of parameter at the beginning
